### DISCUSSION TITLE
Mermaid: subgraph labels are sometimes too dark

### DISCUSSION THREAD

**Discussion by pawamoy** [squidfunk/mkdocs-material#3618] [Q&A]
In some Mermaid diagrams, the labels in subgraphs are too dark when using the slate theme ("openshift" and "service"):
![mermaid1](https://user-images.githubusercontent.com/3999221/155364161-4fd7d697-7a17-403a-9adb-fefd8e4a0d76.png)

But it's weird because in other graphs it's not ("local"):
![mermaid2](https://user-images.githubusercontent.com/3999221/155364218-c9a360b4-77e8-44c2-9db2-c11cb22b6876.png)

Any idea on what is going on 😕 ?

I don't load mermaid.js myself. I don't have any CSS rules touching diagrams.

```
mkdocs==1.2.3
mkdocs-exclude-search==0.6.4
mkdocs-material==8.2.1
mkdocs-material-extensions==1.0.3
mkdocs-spellcheck==0.2.0
```

Here's the code if you want to try it:

````
```mermaid
flowchart LR

%% nodes ----------------------------------------------------------
client["client"]
load_balancer["load balancer"]
api1["API (replica 1)"]
api2["API (replica 2)"]
apin["API (replica N)"]

%% edges ----------------------------------------------------------
client --- load_balancer
subgraph openshift
    subgraph service
        api1 --- api2 --- apin
    end
    load_balancer --- service
end

%% style ----------------------------------------------------------
linkStyle 1 opacity:0
linkStyle 2 opacity:0
```
````

**Comment by facelessuser** [comment]
For comparison, can you post the other diagram source as well?

**Reply by pawamoy** [reply]
Strangely, the one that "works", and I think it's the only one in the whole site, is defined in a Markdown quote that is in turn transformed into an admonition (using markdown-callouts):

````
> NOTE: **How does Copier work?**  
> 
> ```mermaid
> graph TD
> %% nodes ----------------------------------------------------------
> copier("copier")
> gitlab["GitLab"]
> clone("git clone")
> template["/tmp/template"]
> ask_render("ask & render")
> project["project"]
> %% edges ----------------------------------------------------------
> gitlab --> clone
> subgraph local
>     copier --> clone
>     clone --> template
>     copier --> ask_render
>     template --> ask_render
>     ask_render --> project
> end
> %% style ----------------------------------------------------------
> classDef grey fill:#999,stroke:black;
> class clone,ask_render grey;
> ```
> 
> 1. you pass the URL to a template, and a local directory path as arguments
> 1. it clones the template locally, in a temporary folder
> 1. it reads the template questions, then asks these questions to you
> 1. it renders the project from the template, using your answers, creating the directory you specified
> 1. it runs post-generation scripts
````

Screenshots of the HTML output:

Regular diagram:

![image](https://user-images.githubusercontent.com/3999221/155369219-27a4bdc6-8847-4944-9d9a-1d1c0ffb8157.png)

Diagram in quote:

![image](https://user-images.githubusercontent.com/3999221/155369368-2dc200b0-74ce-44c3-9628-198b8740e48a.png)

(sorry about screenshots instead of text, will try to build a minimal reproducible example later)

**Reply by pawamoy** [reply]
CSS of diagram in quotes:

![image](https://user-images.githubusercontent.com/3999221/155370143-fa0a3a1d-7743-486f-937b-852f82b971ed.png)

CSS of the regular diagram:

![image](https://user-images.githubusercontent.com/3999221/155370221-6cab7ae0-fcdd-4581-b2aa-790e06010df5.png)

The first two rules are the problematic ones with the slate theme. Strangely they don't appear on the quoted diagram.

**Reply by facelessuser** [reply]
Yeah, one seems to be a `label` and the other a `cluster-label`. That's why they get styled differently. I don't use Material's Mermaid support currently, but I run into the same issue. I also get some gray bubbles just like Material.

<img width="645" alt="Screen Shot 2022-02-23 at 10 19 30 AM" src="https://user-images.githubusercontent.com/1055125/155372306-f1bf3256-dfd8-4dd8-8481-1eb79035d175.png">

It's definitely something that can be fixed with some styling.

**Reply by facelessuser** [reply]
Personally, the "working" example looks like it should be a `cluster-label` like in the non-working example. Not sure why it's not. Probably just a Mermaid quirk. I wasn't aware of the gray bubble issue either. Not sure what makes one gray and the others not 🤷🏻 .

**Reply by facelessuser** [reply]
Oh, one is a `graph TD` and one is `flowchart LR`. That's why they get constructed differently. they are different diagram types.

**Reply by pawamoy** [reply]
By grey bubbles you mean the "git clone" one for example? It's styled using Mermaid's style options: `classDef grey fill:#999,stroke:black;`. This one I have under control, so it's my responsibility only if it doesn't display well :slightly_smiling_face: But the clusters labels, I was not even able to "override" the CSS :sweat_smile:

**Reply by pawamoy** [reply]
Oh yes, good catch.

**Reply by facelessuser** [reply]
You may be able to control it with:

```css
.cluster text {
    fill: /*my color*/;
}
```

I use the `themeCSS` config to do such things. I don't think I've used `classDef` before. I'm not sure if Material lets you directly change the config or not...again, I'm using my own custom solution.

**Comment by squidfunk** [comment]
Mermaid.js support is pretty new in the community edition, and as it's now available to many more users, I expect that we'll find cases where we need to add some missing styles. I've explored several diagram types and tried to add styles for all cases I could enumerate, but this is clearly a moving target. Furthermore, new versions of Mermaid.js might induce necessary adjustments.

I'm happy to merge PRs. Styles are defined [here](https://github.com/squidfunk/mkdocs-material/blob/master/src/assets/javascripts/components/content/code/mermaid/index.css).

**Reply by pawamoy** [reply]
Thanks @squidfunk! I believe we can simply copy the first flowchart rule into a new "graph" section?

```css
/* ----------------------------------------------------------------------------
 * Rules: graphs
 * ------------------------------------------------------------------------- */

/* Graph cluster label */
.cluster-label {
  color: var(--md-mermaid-label-fg-color);
  font-family: var(--md-mermaid-font-family);
}
```

I'll see if I can setup the project locally to test it. Or maybe it's possible to patch the asset in my own project?

**Reply by squidfunk** [reply]
The style sheets need to be inlined and set via `themeCSS`, so I'm afraid you'll have to set up the theme development process.
